demo/visualizer.js

```diff
@@ -0,0 +1,329 @@
+/*! @license
+ * Shaka Player
+ * Copyright 2016 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+goog.provide('shakaDemo.Visualizer');
+
+goog.require('shakaDemo.BoolInput');
+
+
+/**
+ * Manages a visualizer that shows the buffering progress of the player.
+ */
+shakaDemo.Visualizer = class {
+  /**
+   * @param {!HTMLCanvasElement} canvas
+   * @param {!HTMLElement} div
+   * @param {!HTMLElement} screenshotDiv
+   * @param {!HTMLElement} controlsDiv
+   * @param {!HTMLVideoElement} video
+   * @param {!shaka.Player} player
+   */
+  constructor(canvas, div, screenshotDiv, controlsDiv, video, player) {
+    /** @private {!HTMLCanvasElement} */
+    this.canvas_ = canvas;
+
+    /** @private {!HTMLElement} */
+    this.div_ = div;
+
+    this.active = false;
+
+    /** @private {!HTMLElement} */
+    this.screenshotDiv_ = screenshotDiv;
+
+    /** @private {!HTMLVideoElement} */
+    this.video_ = video;
+
+    /** @private {!shaka.Player} */
+    this.player_ = player;
+
+    /** @private {boolean} */
+    this.takeAutoScreenshots_ = false;
+
+    /** @private {shaka.util.Timer} */
+    this.timer_ = new shaka.util.Timer(() => {
+      this.ageUpdates_();
+      this.updateCanvas_();
+      this.takeAutomaticScreenshots_();
+    });
+
+    /** @private {number} */
+    this.lastCurrentTime_ = 0;
+
+    /** @private {number} */
+    this.colorIOffset_ = 0;
+
+    /** @private {boolean} */
+    this.autoScreenshotTaken_ = false;
+
+    /**
+     * @private {!Array.<{
+     *   age: number,
+     *   start: number,
+     *   end: number,
+     *   isAudio: boolean,
+     * }>}
+     */
+    this.updates_ = [];
+
+    // Listen for when new buffers are appended.
+    player.addEventListener('segmentAppended', (event) => {
+      const start = /** @type {number} */ (event['start']);
+      const end = /** @type {number} */ (event['end']);
+      const contentType = /** @type {string} */ (event['contentType']);
+      let isAudio = false;
+      if (contentType == 'audio') {
+        isAudio = true;
+      } else if (contentType != 'video') {
+        // If it's not video or audio, either it's text or it's invalid.
```

Comment:
why not display the text? If the text is segmented it may be useful to have this information.